### PR TITLE
93 lstm searchspace specification

### PR DESCRIPTION
**Pull request by mokouMonday**
Add Lstm support in Search space.
Pico example code might not work because I couldn't be bothered to change everything multiple times (Please don't always duplicate everything).
Tests took too long so don't know if all of them run through (Please test in a different way than running a nas multiple times over multiple trials).

### CODE REVIEW

File: elasticai/explorer/hw_nas/search_space/architecture_components.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
+from torch import nn
+
+
+class SimpleLSTM(nn.Module):
+    def __init__(self, input_size, hidden_size, num_layers, bidirectional, batch_first):
```
Comment (chrisringhofer) [inline elasticai/explorer/hw_nas/search_space/architecture_components.py]: Pls add dropout :)

File: elasticai/explorer/hw_nas/search_space/layer_builder.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+        self, trial, block, search_params, block_id, input_shape, output_shape
+    ):
+        self.trial = trial
+        self.block = block
+        self.search_params = search_params
+        self.block_id = block_id
+        self.input_shape = input_shape
+        self.output_shape = output_shape
+        self.layers = []
+
```
Comment (chrisringhofer) [inline elasticai/explorer/hw_nas/search_space/layer_builder.py]: Add abstractmethod build

File: elasticai/explorer/hw_nas/search_space/layer_builder.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+    def get_layers(self):
+        return self.layers, self.input_shape
+
+
+@register_layer("linear")
+class LinearBuilder(LayerBuilder):
+    def build(self, num_layers, is_last_block):
+        if isinstance(self.input_shape, list):
+            self.layers.append(nn.Flatten())
```
Comment (chrisringhofer) [inline elasticai/explorer/hw_nas/search_space/layer_builder.py]: Remove in favor of Adapters

File: elasticai/explorer/hw_nas/search_space/layer_builder.py [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+                self.trial,
+                f"num_lstm_layers_b{self.block_id}_l{i}",
+                self.search_params["num_lstm_layers"],
+            )
+            bidirectional = parse_search_param(
+                self.trial,
+                f"bidirectional_b{self.block_id}_l{i}",
+                self.search_params["bidirectional"],
+            )
+
```
Comment (chrisringhofer) [inline elasticai/explorer/hw_nas/search_space/layer_builder.py]: Dropout

**Review by RobinFeldmann — CHANGES_REQUESTED** [review]
I think it looks really good, I like the registry for layers. It makes model building much more adaptable. My only remark is about Typing, except for the nn.Modules all other function signatures should have typehints like in the rest of the explorer. I commented a few of them but there are some more missing typehints.

**Review by chrisringhofer — APPROVED** [review]
All good, all remarks resolved! 👍 

Please add a test and/or issue for handling cases where not every hyperparameter is specified in the search_space.yaml (proposed solution: use default values)
